pkg/ccl/importccl/import_pgdump.go

```diff
@@ -0,0 +1,503 @@
+			ctx.WithReformatTableNames(nil, func() {
+				ctx.FormatNode(tn)
+			})
+		}),
+	)
+	if err != nil {
+		return "", err
+	}
+	f.FormatNode(n)
+	return f.CloseAndGetString(), nil
+}
+
+func bufferDDLPostgresStatement(
+	ctx context.Context,
+	evalCtx *tree.EvalContext,
+	postgresStmt interface{},
+	p sql.JobExecContext,
+	parentID descpb.ID,
+	handler *postgresDDLHandler,
+) error {
+	switch stmt := postgresStmt.(type) {
+	case *tree.CreateDatabase:
+		// If we have previously seen a `CREATE DATABASE` statement then we error
+		// out.
+		if handler.dumpDatabaseName != "" {
+			return errors.Newf("encountered more than one `CREATE DATABASE` statement when importing PGDUMP file")
+		}
+		handler.dumpDatabaseName = string(stmt.Name)
+	case *tree.CreateSchema:
+		// If the schema specifies an explicit database name, replace it with the
+		// temporary pgdump database being imported into.
+		if stmt.Schema.ExplicitCatalog {
+			if stmt.Schema.CatalogName != tree.Name(handler.dumpDatabaseName) {
+				return errors.AssertionFailedf("catalog name %s does not match dump target database name %s",
+					stmt.Schema.CatalogName, handler.dumpDatabaseName)
+			}
+			stmt.Schema.CatalogName = importTempPgdumpDB
+		}
+		handler.bufferedDDLStmts = append(handler.bufferedDDLStmts, formatPostgresStatement(stmt))
+	case *tree.CreateTable:
+		// If the `CREATE TABLE` specifies an explicit database name, replace it
+		// with the temporary pgdump database being imported into.
+		s, err := rewritePostgresStatementTableName(stmt, handler.dumpDatabaseName)
+		if err != nil {
+			return err
+		}
+		handler.bufferedDDLStmts = append(handler.bufferedDDLStmts, s)
+	case *tree.AlterTable:
+		// If the `ALTER TABLE` statement has an explicit database name, replace it
+		// with the temporary pgdump database being imported into.
+		if stmt.Table.HasExplicitCatalog() {
+			if stmt.Table.Parts[2] != handler.dumpDatabaseName {
+				return errors.AssertionFailedf("catalog name %s does not match dump target database name %s",
+					stmt.Table.Parts[2], handler.dumpDatabaseName)
+			}
+			stmt.Table.Parts[2] = importTempPgdumpDB
+		}
+		for _, cmd := range stmt.Cmds {
+			switch cmd := cmd.(type) {
+			case *tree.AlterTableAddConstraint:
+				switch con := cmd.ConstraintDef.(type) {
+				case *tree.ForeignKeyConstraintTableDef:
+					// TODO(adityamaru): handle FKs and fk skip option.
+					if con.Table.ExplicitCatalog {
+						con.Table.CatalogName = importTempPgdumpDB
+					}
+				default:
+					// TODO(adityamaru): confirm that not other constraint can have a
+					// qualified table name.
+				}
+			case *tree.AlterTableSetDefault:
+			case *tree.AlterTableSetVisible:
+			case *tree.AlterTableAddColumn:
+				if cmd.IfNotExists {
+					return wrapErrorWithUnsupportedHint(errors.Errorf("unsupported statement: %s", stmt))
+				}
+			case *tree.AlterTableSetNotNull:
+			default:
+				return wrapErrorWithUnsupportedHint(errors.Errorf("unsupported statement: %s", stmt))
```

Comment:
[not for this PR] I imagine we have all of these cases so we can eventually handle the "ignore_unsupported_statement" flag.  But, I'm wondering if there is a way to make sql deal with what is and isn't supported. For example, if we just sent all statements to SQL and were able to inspect the error to see if it failed because the statement wasn't supported.  This would produce a problem for implementing `ignore_unsupported_statement` (especially if we wanted to issue all of the DDL statements in a single transaction, but perhaps we could do something with savepoints).